### PR TITLE
Upstreaming bevy_color. 

### PR DESCRIPTION
**Pull request by viridia**
# Objective

This provides a new set of color types and operations for Bevy.

Fixes: #10986 #1402 

## Solution

The new crate provides a set of distinct types for various useful color spaces, along with utilities for manipulating and converting colors.

This is not a breaking change, as no Bevy APIs are modified (yet).

### CODE REVIEW

**Comment by viridia** [comment]
@SME-Rendering This has been discussed with cart and others in discord ui-dev.

**Comment by viridia** [comment]
One of the test failures seems to be platform-specific. I'm using the existing Bevy functions for converting to and from Lcha, and as you can see it's failing on windows but not linux or mac. And this isn't a small failure either, it's like '0 degrees != 90 degrees'.

The way the conversion tests work is that there's a generated test file that uses the `palette` crate as the source of truth.

**Comment by alice-i-cecile** [comment]
@bushrat011899 can I get your review / opinions on this, based on your work in https://github.com/bevyengine/bevy/pull/9698 ?

**Comment by viridia** [comment]
@bushrat011899 I recognize that you've done a lot of work on this, more than I have. The main difference here is that my version doesn't have any external dependencies. It does use the `palette` crate to generate test data, but the actual library does not depend on `palette`.

This library can and should be extended with more features, many of which are present in your PR. However, I wanted to make this small enough to be digestible to reviewers.

**Review by alice-i-cecile — COMMENTED** [review]
Does the `gen_tests` crate make more sense in our existing [`tools` directory](https://github.com/bevyengine/bevy/tree/main/tools)?

**Comment by viridia** [comment]
> Does the `gen_tests` crate make more sense in our existing [`tools` directory](https://github.com/bevyengine/bevy/tree/main/tools)?

Maybe. It's an extremely specialized tool that is only useful in this one context.

File: crates/bevy_color/src/color_ops.rs
Comment (alice-i-cecile) [inline crates/bevy_color/src/color_ops.rs:9]: I don't think `alpha` is the right variable name.

File: crates/bevy_color/src/color_ops.rs
Comment (alice-i-cecile) [inline crates/bevy_color/src/color_ops.rs:8]: This should specify what happens if values outside of the range are supplied (and what the sensible range is).

**Comment by viridia** [comment]
There are three test failures. One is not my fault (I think). One is a doc problem which I can fix. The third I'm puzzled by, it's a disparity between what the bevy_render LCH conversion and the palette LCH conversion, the hue component comes out different. It's troubling that the test fails on windows only.

That particular test is kind of funky anyway, I had to give a fairly large tolerance (1.7 degrees) to get it to pass on any platform. I hesitate to point the finger at bevy_render, but I trust palette's answer more. Either that, or my test data generator is calling things incorrectly.

File: crates/bevy_color/src/color_ops.rs
Comment (alice-i-cecile) [inline crates/bevy_color/src/color_ops.rs:20]: This is a distributive property, not commutative.

File: crates/bevy_color/src/color_ops.rs
Comment (alice-i-cecile) [inline crates/bevy_color/src/color_ops.rs:11]: We should mention what happens when overly dark or bright values would be returned.

File: crates/bevy_color/src/color_ops.rs
Comment (alice-i-cecile) [inline crates/bevy_color/src/color_ops.rs:40]: This should probably have a `alpha` method to get the alpha as well.

File: crates/bevy_color/src/color_range.rs
Comment (alice-i-cecile) [inline crates/bevy_color/src/color_range.rs:9]: Why can't we just make these fields pub?

File: crates/bevy_color/src/color_ops.rs
Comment (alice-i-cecile) [inline crates/bevy_color/src/color_ops.rs:27]: Organizationally, it feels like this trait should be stored together with `ColorRange` in a `color_mixing.rs` file.

File: crates/bevy_color/src/color_range.rs
Comment (alice-i-cecile) [inline crates/bevy_color/src/color_range.rs:29]: The docs aren't sufficiently clear on why this is needed, how it can be used, or how it relates to `ColorRange`.

File: crates/bevy_color/src/hsla.rs
Comment (alice-i-cecile) [inline crates/bevy_color/src/hsla.rs:6]: Wikipedia links for these color space types would be a nice addition.

File: crates/bevy_color/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_color/src/lib.rs:14]: We should add a short note here about how you should use `Okalba` or at least `Hsla` for most color blending operations.

File: crates/bevy_color/src/testing.rs
Comment (alice-i-cecile) [inline crates/bevy_color/src/testing.rs:1]: We might be able to replace these allows with a test cfg instead?

File: crates/bevy_color/src/color_difference.rs
Comment (alice-i-cecile) [inline crates/bevy_color/src/color_difference.rs:5]: Definitely feels like a `bevy_math` trait that ended up here 😅 I'm okay to keep this in here for now, but we may want to think about how to represent / expose this later.

File: crates/bevy_color/src/oklaba.rs
Comment (alice-i-cecile) [inline crates/bevy_color/src/oklaba.rs:32]: The l/c/h here don't match the l/a/b below.

File: crates/bevy_color/src/color_ops.rs
Comment (msvbg) [reply crates/bevy_color/src/color_ops.rs:20]: I don't think distributivity is fully correct either. Perhaps it's best to just leave out the terminology?

File: crates/bevy_color/src/lib.rs
Comment (bushrat011899) [inline crates/bevy_color/src/lib.rs:9]:
```suggestion
//! - [`Oklaba`] (lightness, a-axis, b-axis, alpha)
```

File: crates/bevy_color/src/linear_rgba.rs
Comment (bushrat011899) [inline crates/bevy_color/src/linear_rgba.rs:24]:
```suggestion
    /// Construct a new [`LinearRgba`] color from components.
```

File: crates/bevy_color/src/linear_rgba.rs
Comment (bushrat011899) [inline crates/bevy_color/src/linear_rgba.rs:192]:
```suggestion
        // From https://github.com/Ogeon/palette/blob/e75eab2fb21af579353f51f6229a510d0d50a311/palette/src/oklab.rs#L312-L332
```

I'm not a copyright expert by any means, but do we need to attribute any credit to `palette` for this? E.g., in the `README.md`

File: crates/bevy_color/src/srgba.rs
Comment (bushrat011899) [inline crates/bevy_color/src/srgba.rs:23]: I think it would be good to include a link to where these definitions come from? E.g., [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/named-color)

**Review by bushrat011899 — COMMENTED** [review]
This looks great! I can appreciate the desire to keep 3rd parties to a minimum, and I think bringing in `palette` as a comparison for testing/validation is a really good middle ground. I've probably been very nit-picky on some of the suggestions which I'll just apologise for upfront!

I would like to expand this to include some more colour spaces (XYZ, Oklch, etc.) but those are really well suited for follow-up PRs so I'm happy with the current amount of of spaces supported.

Probably the two biggest bits of feedback I have are:

1. I think the color operations should take `self` and produce `Self` to allow the space itself to decide if `Copy` is appropriate, and default to move semantics.
2. I think including CIEDE2000 perceptual difference is important enough to consider for this initial PR, but I also understand moving that into a followup.

File: crates/bevy_color/src/color_ops.rs
Comment (bushrat011899) [reply crates/bevy_color/src/color_ops.rs:20]: I believe the correct term is "linear", since the property is `f(a + b) = f(a) + f(b)`, which is a specialised case of the definition of a linear operation `f(ax + b) = a*f(x) + f(b)`

File: crates/bevy_color/src/color_ops.rs
Comment (alice-i-cecile) [inline crates/bevy_color/src/color_ops.rs:27]: Blocking concern: we need to figure out how this trait relates to `Animatable`. Our options:

1. Use and implement `Animatable` directly, and axe this trait.
2. Use a blanket `impl <T: Mix> Animatable for T` that acts as a bridging layer.

I think that 1 is clearer and results in less duplication, but will complicate our dependency tree and make this crate effectively useless externally. As a result, I think that 2 is the way to go.

File: crates/bevy_color/Cargo.toml
Comment (JMS55) [inline crates/bevy_color/Cargo.toml:12]: Needs removing

File: crates/bevy_color/src/color_representation.rs
Comment (JMS55) [reply crates/bevy_color/src/color_representation.rs:8]: Any reason not to just call it `Color` like the old API, if this is meant to be the user-facing type we use in all of our components?

Comment (JMS55) [inline crates/bevy_color/src/color_representation.rs:8]: I'd like to put some docs/constraints here on whether the alpha-colors are premultiplied or separate alpha.

File: crates/bevy_color/src/lib.rs
Comment (JMS55) [reply crates/bevy_color/src/lib.rs:14]: To add to this, I'd like the following:
* Short comparisons between the color spaces with preferred defaults for different use cases, including when to use ColorRepresentation.
* Links to online color pickers for each color space (maybe, up to maintainers)
* Explanation of color space, gamut, display ranges, and other terms
* Some links to references on how to understand color in more detail

**Review by JMS55 — CHANGES_REQUESTED** [review]
Not a thorough review, but did a quick skim. My main request is more docs to help people understand the nuances of colors, what Display P3 is vs okLCH, LDR vs HDR, etc. I'd write some suggestions but I don't understand myself :P.

File: crates/bevy_color/src/color_ops.rs
Comment (JMS55) [reply crates/bevy_color/src/color_ops.rs:27]: 3. Move the stuff to a wrapper type, e.g. `PerceputallyLinearMix<T>`. Or differentiate perceptually vs mathematically linear by the underlying color space types.

File: crates/bevy_color/src/lib.rs
Comment (bushrat011899) [reply crates/bevy_color/src/lib.rs:14]: >     * Links to online color pickers for each color space (maybe, up to maintainers)

Definitely an extension beyond the initial PR, but this sounds like an excellent case for a Bevy tool to be provided, perhaps in the `examples` folder. I think there's been quite a few examples of colour pickers made using the current Bevy UI system, and a colour picking tool in Bevy is a nice demonstration of how to use `bevy_color` _and_ an explanation of the broader concept of colours as well.

File: crates/bevy_color/src/color_representation.rs
Comment (bushrat011899) [reply crates/bevy_color/src/color_representation.rs:8]: If this becomes the canonical Bevy colour type I think taking the name `Color` is reasonable.

File: crates/bevy_color/src/srgba.rs
Comment (viridia) [reply crates/bevy_color/src/srgba.rs:23]: I copied them from the bevy Color class. The colors are a subset of the ones listed on MDN.

Having these is handy for debugging, but I don't know whether we should spend too much effort to ensure that the set of colors matches some external standard.

File: crates/bevy_color/src/testing.rs
Comment (viridia) [reply crates/bevy_color/src/testing.rs]: Done

File: crates/bevy_color/src/color_difference.rs
Comment (viridia) [reply crates/bevy_color/src/color_difference.rs:5]: ok

File: crates/bevy_color/src/color_range.rs
Comment (viridia) [reply crates/bevy_color/src/color_range.rs:22]: Expanded comment.

File: crates/bevy_color/src/lib.rs
Comment (viridia) [reply crates/bevy_color/src/lib.rs:14]: I've expanded the explanation of color spaces, but not the other suggestions (gamut etc.). Mainly because I'm not confident that my explanation of these would be correct.

File: crates/bevy_color/src/color_ops.rs
Comment (alice-i-cecile) [reply crates/bevy_color/src/color_ops.rs:44]: It's very useful when working with trait objects.

Comment (viridia) [reply crates/bevy_color/src/color_ops.rs:44]: Done. (Will be in next push).

File: crates/bevy_color/src/lib.rs
Comment (JMS55) [reply crates/bevy_color/src/lib.rs:14]: Linking to other resources would be fine

File: crates/bevy_color/src/color_representation.rs
Comment (JMS55) [reply crates/bevy_color/src/color_representation.rs:8]: We should figure out what variant we _do_ use (assuming we only use one system), and document that these color types operate under that imo.

File: crates/bevy_color/src/srgba.rs
Comment (viridia) [reply crates/bevy_color/src/srgba.rs:23]: I've edited the list of colors so that it now only includes the 16 VGA colors + NONE + DARK_GRAY.

File: crates/bevy_color/src/srgba.rs
Comment (afonsolage) [inline crates/bevy_color/src/srgba.rs:117]: Why not use the built in `from_str_radix` then just do some bitwise operation on either u16 or u32 to convert back to RGBA bytes?

I think using `from_str_radix` and some form of bitwise will be faster due to cache friendness and no loops.

**Review by afonsolage — COMMENTED** [review]
I like the idea of having `bevy_color` and I love it has a focus on `const`  fns too.

File: crates/bevy_color/src/linear_rgba.rs
Comment (viridia) [reply crates/bevy_color/src/linear_rgba.rs:72]: Done

File: crates/bevy_color/src/srgba.rs
Comment (viridia) [reply crates/bevy_color/src/srgba.rs:117]: Bear in mind that this needs to work with 3, 4, 6 and 8 digits.

I think I will leave this for now. "First make it correct, then make it fast". If someone wants to make an optimized version, well, there's lots of unit tests to validate.

**Review by afonsolage — APPROVED** [review]

File: crates/bevy_color/src/hsla.rs
Comment (bushrat011899) [inline crates/bevy_color/src/hsla.rs:52]: Possible consistency issue: some colours default to white (`Color`, `Srgba`, `LinearRgba`), while others default to black (`Hsla`, `Lcha`, `Oklaba`). I don't mind which is the default, but it should probably be consistent.

**Review by bushrat011899 — APPROVED** [review]
Looks good! Just a minor note regarding consistent defaults, otherwise I think this is good to go!

File: crates/bevy_color/src/color_range.rs
Comment (alice-i-cecile) [inline crates/bevy_color/src/color_range.rs:27]: I'm still having difficulty piecing together the usage of this type from the definition and docs alone. Can you add a doc test that shows how this trait could be used for interpolating for e.g. a gradient, and why this trait is needed?

**Review by alice-i-cecile — APPROVED** [review]
I'm happy enough with this in its current form to merge this and iterate in follow-ups, provided that @cart thinks that's the right path from a workflow perspective.

I'm personally weakly in favor of that: I think we *can* solve any remaining issues in follow-ups, and it will be easier to review and debate across PRs.

Outstanding requests:

1. Better docs on each color space.
2. Clearer usage examples for `ColorRange` and `AnyColorRange`.
3. Integration with the rest of Bevy and removal of existing color code.
4. Interplay between `Animatable`, `ColorRange`, `AnyColorRange` and `Mix`. It feels like there's too many related traits, and their interplay isn't clear to either contributors or users.
5. Fluent-style helper methods like `with_r`.

File: crates/bevy_color/src/color_range.rs
Comment (viridia) [reply crates/bevy_color/src/color_range.rs:27]: I can also just remove this. I use it in my own UI work, but I can just as easily move the trait definition to my UI crate.

The way I use this is that I have an ECS component that animates a UI property: examples are `AnimatedBackgroundColor`, `AnimatedBorderColor`, `AnimatedOutlineColor` and so on. Within the component there's a color range that specifies the beginning and ending of the transition. You can think of the start and end as keyframes in an animation sequence, except that there's only two.

The `AnyColorRange` trait lets me interpolate in different color spaces without having to have different component types, i.e. `AnimatedBackgroundColorRgb`, `AnimatedBackgroundColorHsl` and so on. Because it's a dyn trait, the color space type is erased.

The problem with adding a doc test is that the use case is rather elaborate, and it would make for a fairly verbose test.

Comment (alice-i-cecile) [reply crates/bevy_color/src/color_range.rs:27]: Yeah, let's cut this for now: we can always add it back in once we have a tangible first-party use case.

File: crates/bevy_color/src/hsla.rs
Comment (viridia) [reply crates/bevy_color/src/hsla.rs]: Done

File: crates/bevy_color/src/color_ops.rs
Comment (msvbg) [reply crates/bevy_color/src/color_ops.rs]: At the risk of dangerous levels of bikeshedding... Rewriting `lighter` to $\ell$ you might phrase it as $\ell(b) \circ \ell(a) = \ell(a + b)$ which could be called something like additivity or additivity under composition.

File: crates/bevy_color/Cargo.toml
Comment (mockersf) [inline crates/bevy_color/Cargo.toml:16]:
```suggestion
bevy_math = { path = "../bevy_math", version = "0.14.0-dev" }
bevy_reflect = { path = "../bevy_reflect", version = "0.14.0-dev", features = [
  "bevy",
] }
bevy_render = { path = "../bevy_render", version = "0.14.0-dev" }
```

**Comment by viridia** [comment]
> Outstanding requests:

> 1. Better docs on each color space.

I've done a preliminary search on this, and unfortunately most of the web pages that explain color are lengthy treatises on color theory, human perception and device capabilities. There's no obvious place you can link to directly that gives you a concise definition of "LHA" color space for example.

> 2. ~Clearer usage examples for `ColorRange` and `AnyColorRange`.~
> 3. Integration with the rest of Bevy and removal of existing color code.

My current thinking is that rather than changing all of the Bevy APIs for 0.14, we actually refine and improve bevy_color over the next release cycle before doing the integration. This means that people who use bevy_color will need to call `.into()` when passing the color to ECS components like `StandardMaterial` or `BorderColor`. If there are places where we can accept an `Into<Color>` and be backwards-compatible, those can be done right away.

Also, we need to decide whether the fundamental color type used in StandardMaterial is eventually going to be `Color` or `LinearRgba`. It would be nice to avoid the runtime check of the enum when doing low-level setting of uniforms.

> 4. Interplay between `Animatable`, `ColorRange`, ~`AnyColorRange`~ and `Mix`. It feels like there's too many related traits, and their interplay isn't clear to either contributors or users.

I don't see these as overlapping as much as you do. `Mix` is used for other things than just time-based interpolation - for example, relative (vs absolute) lightening and darkening is something you do with `Mix`.

> 5. Fluent-style helper methods like `with_r`.

Sure. I don't want to go overboard with helper methods, but I'm ok with adding the most commonly used fluent methods. Bear in mind that you can get the same effect, slightly more verbosely, with:

```rust
let new_color = Rgba { red: new_red, ...old_color };
```
And in fact, this is actually more flexible if you're setting more than one channel:
```rust
let new_color = Rgba { red: new_red, green: new_green, ...old_color };
```
Similarly, the automatic conversion of the enum type (`Color`) can be done like this:
```rust
let new_color = Rgba { red: new_red, ...old_hsl_color.into() };
```
So basically, all of the things that you might want to do with the fluent methods can already be done using Rust idioms; it's just that those idioms are not quite as obvious.

**Comment by ameknite** [comment]
I think this should also add bevy_color to [publish.sh](https://github.com/bevyengine/bevy/blob/main/tools/publish.sh)

**Comment by JMS55** [comment]
> I've done a preliminary search on this, and unfortunately most of the web pages that explain color are lengthy treatises on color theory, human perception and device capabilities. There's no obvious place you can link to directly that gives you a concise definition of "LHA" color space for example.

I'm fine linking to those in the module docs in lieu of more specific docs.

**Comment by viridia** [comment]
I've added a general link to the Wikipedia article on color spaces in the crate-level doc. I also added a bit more detail on color transformations.

File: crates/bevy_color/src/lib.rs
Comment (cart) [inline crates/bevy_color/src/lib.rs:12]: This line is duplicated

File: crates/bevy_color/src/lib.rs
Comment (cart) [inline crates/bevy_color/src/lib.rs:23]: I don't think GPU hardware _specifically_ cares about linear rgb vs srgb. It will happily operate on either representation. Our rendering code requires linear rgb because that is what it expects. We _could_ write an srgb renderer, it would just suffer from that choice. The expected output format for the window surface / swapchain is platform specific and often it can be configured to use either srgb or linear formats.

File: crates/bevy_color/src/color.rs
Comment (cart) [inline crates/bevy_color/src/color.rs:21]: I suspect if we swap our current Color type for this, that we will want to add a bunch of convenience constructors here. But that can wait!

File: crates/bevy_color/Cargo.toml
Comment (cart) [inline crates/bevy_color/Cargo.toml:17]: Specifying the version at the patch precision doesn't do anything functionally and actually makes it harder to keep everything in sync via find/replace. We currently specify serde "1.0" or "1" everywhere else (I prefer "1.0").

**Review by cart — APPROVED** [review]
Looks good to me! I'm reasonably confident that this is the right path forward. I'm down to merge this provided we still consider it in the "candidate for upstreaming" phase. I don't think we can fully evaluate its fitness for Bevy until it has been ported.

**Comment by cart** [comment]
> Sure. I don't want to go overboard with helper methods, but I'm ok with adding the most commonly used fluent methods. Bear in mind that you can get the same effect, slightly more verbosely, with:

I'm in a similar camp. I like Bevy to be as "field driven" as possible for most types, especially the high-traffic ones. Builder-style should be a last-resort as it increases API surface. I like leaning on Rust as much as possible.

**Comment by cart** [comment]
> My current thinking is that rather than changing all of the Bevy APIs for 0.14, we actually refine and improve bevy_color over the next release cycle before doing the integration. This means that people who use bevy_color will need to call .into() when passing the color to ECS components like StandardMaterial or BorderColor. If there are places where we can accept an Into<Color> and be backwards-compatible, those can be done right away.

I think it will be hard to refine this further without adopting it internally / we risk refining in the wrong direction. I'd prefer a hard switch PR asap where we can tweak / polish / evaluate the new approach. I think a "middle ground / support both" release would just be confusing and introduce more churn.

> Also, we need to decide whether the fundamental color type used in StandardMaterial is eventually going to be Color or LinearRgba. It would be nice to avoid the runtime check of the enum when doing low-level setting of uniforms.

Agreed. I think this is the "biggest" / most important choice to be made. This also applies to other areas, such as UI / sprite color. When a user (or internal Bevy dev) needs a color in their component, what do they reach for? We should be consistent and make doing the "recommended" thing easy by giving "good names" to the recommended types (ex: we should probably rename `enum Color` to something else if we decide that Srgba should be used in most places, and maybe even consider renaming Srgba to Color).

File: crates/bevy_color/src/lib.rs
Comment (viridia) [reply crates/bevy_color/src/lib.rs]: Done

File: crates/bevy_color/src/lib.rs
Comment (viridia) [reply crates/bevy_color/src/lib.rs]: Comment updated.

File: crates/bevy_color/Cargo.toml
Comment (viridia) [reply crates/bevy_color/Cargo.toml]: Done.

**Comment by JMS55** [comment]
For deciding Color vs LinearRgba for StandardMaterial, reflection and editor-driven-development are important considerations.

**Comment by ameknite** [comment]
if this fixes https://github.com/bevyengine/bevy/issues/1402 why is not closed?

**Pull request merged**
